src/libstore/indirect-root-store.hh

```diff
@@ -0,0 +1,71 @@
+#pragma once
+///@file
+
+#include "local-fs-store.hh"
+
+namespace nix {
+
+/**
+ * Mix-in class for implementing permanent roots as a pair of a direct
+ * (strong) reference and indirect weak reference to the first
+ * reference.
+ *
+ * See methods for details on the operations it represents.
+ *
+ * @note
+ * To understand the purpose of this class, it might help to do some
+ * "closed-world" rather than "open-world" reasoning, and consider the
+ * problem it solved for us. This class was factor out from
+ * `LocalFSStore` in order to support the following table, which
+ * contains 4 concrete store types (non-abstract classes, exposed to the
+ * user), and how they implemented the two GC root methods:
+ *
+ * @note
+ * |                   | `addPermRoot()` | `addIndirectRoot()` |
+ * |-------------------|-----------------|---------------------|
+ * | `LocalStore`      | local           | local               |
+ * | `UDSRemoteStore`  | local           | remote              |
+ * | `SSHStore`        | doesn't have    | doesn't have        |
+ *
+ * @note
+ * Note how only the local implementations of `addPermRoot()` need
+ * `addIndirectRoot()`; that is what this class enforces. Without it,
+ * and with `addPermRoot()` and `addIndirectRoot()` both `virtual`, we
+ * would accidentally be allowing for a combinatorial explosion of
+ * possible implementations many of which make no sense. Having this and
+ * that invariant enforced cuts down that space.
+ */
+struct IndirectRootStore : public virtual LocalFSStore
```

Comment:
I think that name is not great. The most important bit here isn't that the store can register indirect roots (any `ssh-ng` could in theory), but that it can register permanent roots (although `addPermRoot` is implemented and `final` here, it only makes sense for stores that have a local filesystem baking)

Reply:
The important bit is in fact the indirect roots.

> it only makes sense for stores that have a local filesystem baking

It is a subclass of `LocalFSStore`, so unless you meant something like "non ssh/NFS" by "local" that is in fact always the case.

`LocalFSStore` defines `addPermRoot`. So the point of this class isn't to define that interfaec, but to merely implement it. The point is the indirect root implementation.